tools/text/msgdis.py

```diff
@@ -0,0 +1,3190 @@
+    0x617C: "NA_SE_VO_TIVO05",
+    0x617D: "NA_SE_VO_OMVO00",
+    0x617E: "NA_SE_VO_OMVO01",
+    0x617F: "NA_SE_VO_OMVO02",
+    0x6180: "NA_SE_VO_OMVO03",
+    0x6181: "NA_SE_VO_OMVO04",
+    0x6182: "NA_SE_VO_OMVO05",
+    0x6183: "NA_SE_VO_OMVO06",
+    0x6184: "NA_SE_VO_DEMO_FALL2",
+    0x6185: "NA_SE_VO_DEMO_384",
+    0x6186: "NA_SE_VO_DEMO_385",
+    0x6187: "NA_SE_VO_DEMO_386",
+    0x6188: "NA_SE_VO_DEMO_387",
+    0x6189: "NA_SE_VO_DEMO_388",
+    0x618A: "NA_SE_VO_DEMO_389",
+    0x618B: "NA_SE_VO_DEMO_390",
+    0x618C: "NA_SE_VO_DEMO_391",
+    0x618D: "NA_SE_VO_DEMO_392",
+    0x618E: "NA_SE_VO_DEMO_393",
+    0x618F: "NA_SE_VO_DEMO_394",
+}
+
+def read4(data : bytes, p : int) -> int:
+    return struct.unpack(">I", data[p:p+4])[0]
+
+def unique_or_none(lst : List[T]) -> Optional[T]:
+    if not lst:
+        return None
+    elem = lst[0]
+    for e in lst[1:]:
+        if e != elem:
+            return None
+    return elem
+
+class MessageDecoder:
+    def __init__(self, control_end : int, control_codes : Dict[int, Tuple[str, str, Optional[Tuple[Callable[[int], str]]]]], 
+                 control_header : Tuple[str, str, Optional[Tuple[Callable[[int], str]]]], extraction_charmap : Dict[int, str]) -> None:
+        self.control_end : int = control_end
+        self.control_codes : Dict[int, Tuple[str, str, Optional[Tuple[Callable[[int], str]]]]] = control_codes
+        self.control_header : Tuple[str, str, Optional[Tuple[Callable[[int], str]]]] = control_header
+        self.extraction_charmap : Dict[int, str] = extraction_charmap
+        self.msg : Optional[bytes] = None
+
+    def pop_char(self) -> int:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def pop_char_end(self) -> int:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def decode_char(self, c : int) -> str:
+        # Implement in subclass
+        raise NotImplementedError()
+
+    def pop_byte(self) -> int:
+        c, self.msg = self.msg[0], self.msg[1:]
+        return c
+
+    def pop_byte_end(self) -> int:
+        c, self.msg = self.msg[-1], self.msg[:-1]
+        return c
+
+    def pop_2byte(self) -> int:
+        u = self.pop_byte()
+        l = self.pop_byte()
+        return (u << 8) | l
+
+    def pop_2byte_end(self) -> int:
+        l = self.pop_byte_end()
+        u = self.pop_byte_end()
+        return (u << 8) | l
+
+    def format_sfx_id(self, c : int) -> str:
+        try:
+            if c & 0x800:
+                return sfx_ids[c & ~0x800]
+            else:
+                return f"{sfx_ids[c]} - SFX_FLAG"
+        except KeyError:
+            return f"0x{c:04X}"
+        
+    def format_byte_hex(self, c: int) -> str:
+        return f"0x{c:02X}"
+    
+    def format_2byte_hex(self, c: int) -> str:
+        return f"0x{c:04X}"
+
+    def format_item_id(self, c : int) -> str:
+        try:
+            return item_ids[c]
+        except KeyError:
+            return f"0x{c:04X}"
```

Comment:
Is this try needed? In which cases? Item ids are also encoded as bytes so the format should be `:02X`?

Reply:
It is not. Nothing actually uses this yet. Credits have a ctrl code that allow its use, but nothing in the credits use it.
Normal messages have something in the Header that is almost the itemId, but has a layer of indirection.  I think I added it just because I saw it under the sfx which I needed to add it. I'll remove the try block.